### PR TITLE
keycloak: Add fuzzer for keycloak

### PR DESCRIPTION
**Pull request by arthurscchan**
Add 3 additional fuzzers for keycloak for the providers in the crypto package.

### CODE REVIEW

File: projects/keycloak/CertificateUtilsProviderFuzzer.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+          break;
+        case 5:
+          keyPair = KeyUtils.generateRsaKeyPair(2048);
+          Date startDate = new Date(data.consumeLong());
+          Date expiryDate = new Date(data.consumeLong());
+          provider.createServicesTestCertificate(data.consumeString(data.remainingBytes() / 2),
+              startDate, expiryDate, keyPair, data.consumeRemainingAsString());
+          break;
+      }
+    } catch (Exception e) {
```
Comment (AdamKorcz) [inline projects/keycloak/CertificateUtilsProviderFuzzer.java:88]: Please add a comment why we are catching this broad of an exception.

File: projects/keycloak/ECDSACryptoProviderFuzzer.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+      // Randomly call method from the ECDSACryptoProvider instance
+      Integer length = data.consumeInt();
+      byte[] bytes = data.consumeRemainingAsBytes();
+      if (data.consumeBoolean()) {
+        provider.concatenatedRSToASN1DER(bytes, length);
+      } else {
+        provider.asn1derToConcatenatedRS(bytes, length);
+      }
+    } catch (IOException | NullPointerException | IllegalArgumentException e) {
```
Comment (AdamKorcz) [inline projects/keycloak/ECDSACryptoProviderFuzzer.java:53]: Why are we catching a `NullPointerException` here?

File: projects/keycloak/JweAlgorithmProviderFuzzer.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+          break;
+        case 6:
+          keyPair = generator.generateKeyPair();
+          encryptionKey = keyPair.getPublic();
+          decryptionKey = keyPair.getPrivate();
+          algorithmProvider = new FIPSRsaKeyEncryptionJWEAlgorithmProvider(null);
+          break;
+      }
+
+      // Randomly call method from the JWE algorithm provider instance
```
Comment (AdamKorcz) [inline projects/keycloak/JweAlgorithmProviderFuzzer.java:94]: Please specify in the comment that the fuzzer either encodes or decodes.

File: projects/keycloak/JweAlgorithmProviderFuzzer.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        JWEKeyStorage storage = new JWEKeyStorage();
+        storage.setEncryptionProvider(provider);
+        storage.setEncryptionKey(encryptionKey);
+        storage.setDecryptionKey(decryptionKey);
+
+        algorithmProvider.encodeCek(provider, storage, encryptionKey);
+      } else {
+        algorithmProvider.decodeCek(data.consumeRemainingAsBytes(), decryptionKey);
+      }
+    } catch (Exception e) {
```
Comment (AdamKorcz) [inline projects/keycloak/JweAlgorithmProviderFuzzer.java:106]: Please add a comment why we are catching this broad of an exception.

File: projects/keycloak/build.sh [hunk 5]
```diff
@@ -18,21 +18,33 @@
...
+PATH=$JAVA_HOME/bin:$PATH
 
 # Build Keycloak
 MAVEN_ARGS="-Djavac.src.version=17 -Djavac.target.version=17 "
 MAVEN_ARGS=$MAVEN_ARGS"-DskipTests -Dgpg.skip -Dmaven.source.skip "
 MAVEN_ARGS=$MAVEN_ARGS"-DskipExamples -DskipTestsuite"
 $MVN clean package $MAVEN_ARGS
 
+# Dependency
+wget https://repo1.maven.org/maven2/org/bouncycastle/bc-fips/1.0.2.3/bc-fips-1.0.2.3.jar
```
Comment (AdamKorcz) [inline projects/keycloak/build.sh:32]: Please specify by way of commenting which fuzzers each of these were initially added for.

File: projects/keycloak/build.sh [hunk 6]
```diff
@@ -18,21 +18,33 @@
...
+
 RUNTIME_CLASSPATH=
 
-for JARFILE in $(find ./ -name *.jar)
+for JARFILE in $(find ./ -name "*.jar")
 do
   if [[ "$JARFILE" == *"core/"* ]] || [[ "$JARFILE" == *"saml-core/"* ]] || \
   [[ "$JARFILE" == *"saml-core-api/"* ]] || [[ "$JARFILE" == *"common/"* ]] || \
-  [[ "$JARFILE" == *"jboss-log"* ]] || [[ "$JARFILE" == *"jackson"* ]]
+  [[ "$JARFILE" == *"adapters/"* ]] || [[ "$JARFILE" == *"common/"* ]] || \
```
Comment (AdamKorcz) [inline projects/keycloak/build.sh:42]: Since this is getting excessive, please explain this by way of comments: What is the build script doing here?

**Review by AdamKorcz — CHANGES_REQUESTED** [review]

File: projects/keycloak/JweAlgorithmProviderFuzzer.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+public class JweAlgorithmProviderFuzzer {
+  public static void fuzzerTestOneInput(FuzzedDataProvider data) {
+    try {
+      // Set up a list of valid encryption algorithm for the JWE object
+      String[] enc = {JWEConstants.A128GCM, JWEConstants.A192GCM, JWEConstants.A256GCM};
+
+      // Initialize the base object for key management and generation
+      KeyGenerator keyGenerator = KeyGenerator.getInstance("AES");
+      KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA");
+      generator.initialize(2048);
```
Comment (AdamKorcz) [inline projects/keycloak/JweAlgorithmProviderFuzzer.java:47]: Does the fuzzer need to do this on every iteration? Or can it be taken out of the fuzzer body?

File: projects/keycloak/JweAlgorithmProviderFuzzer.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+      JWEAlgorithmProvider algorithmProvider = null;
+      KeyPair keyPair = null;
+      Key encryptionKey = null;
+      Key decryptionKey = null;
+
+      // Randomly create an JWE Algorithm Provider instance
+      // from different implementation
+      switch (data.consumeInt(1, 6)) {
+        case 1:
+          keyGenerator.init(256);
```
Comment (AdamKorcz) [inline projects/keycloak/JweAlgorithmProviderFuzzer.java:58]: Does the fuzzer need to create a provider on every iteration? Or can it be taken out of the fuzzer body?

**Review by AdamKorcz — CHANGES_REQUESTED** [review]

File: projects/keycloak/CertificateUtilsProviderFuzzer.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    // Initialise base certificate related object
+    CertificateUtilsProvider provider = null;
+    X509Certificate cert = null;
+    KeyPair keyPair = null;
+
+    try {
+      // Randomly create a certificate utils provider instance
+      switch (data.consumeInt()) {
+        case 1:
+          provider = new BCCertificateUtilsProvider();
```
Comment (AdamKorcz) [inline projects/keycloak/CertificateUtilsProviderFuzzer.java:44]: Does the fuzzer need to generate a provider on every iteration, or can a global provider be instantiated and reused?

File: projects/keycloak/ECDSACryptoProviderFuzzer.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        case 3:
+          provider = new BCFIPSECDSACryptoProvider();
+          break;
+      }
+
+      // Randomly call method from the ECDSACryptoProvider instance
+      Integer length = data.consumeInt();
+      byte[] bytes = data.consumeRemainingAsBytes();
+      if (data.consumeBoolean()) {
+        provider.concatenatedRSToASN1DER(bytes, length);
```
Comment (AdamKorcz) [inline projects/keycloak/ECDSACryptoProviderFuzzer.java:49]: Does length of bytes and length need to be identical here? If yes, did you make it variable on purpose?

File: projects/keycloak/JweAlgorithmProviderFuzzer.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+/**
+ * This fuzzer targets the methods in different
+ * JweAlgorithm Provider implementation classes
+ * in the crypto package.
+ */
+public class JweAlgorithmProviderFuzzer {
+  public static void fuzzerTestOneInput(FuzzedDataProvider data) {
+    try {
+      // Set up a list of valid encryption algorithm for the JWE object
+      String[] enc = {JWEConstants.A128GCM, JWEConstants.A192GCM, JWEConstants.A256GCM};
```
Comment (AdamKorcz) [inline projects/keycloak/JweAlgorithmProviderFuzzer.java:42]: Please make this a global var.

**Review by AdamKorcz — CHANGES_REQUESTED** [review]

File: projects/keycloak/ECDSACryptoProviderFuzzer.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+ * in the crypto package.
+ */
+public class ECDSACryptoProviderFuzzer {
+  public static void fuzzerTestOneInput(FuzzedDataProvider data) {
+    try {
+      // Randomly create an instance of ECDSACryptoProvider
+      ECDSACryptoProvider provider = null;
+      switch (data.consumeInt(1, 3)) {
+        case 1:
+          provider = new BCECDSACryptoProvider();
```
Comment (AdamKorcz) [inline projects/keycloak/ECDSACryptoProviderFuzzer.java:35]: Does the fuzzer need to create a new provider on every iteration, or can it be taken out of the fuzzers body and reused when needed?

File: projects/keycloak/JweAlgorithmProviderFuzzer.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+          keyPair = generator.generateKeyPair();
+          encryptionKey = keyPair.getPublic();
+          decryptionKey = keyPair.getPrivate();
+          algorithmProvider = new FIPSRsaKeyEncryptionJWEAlgorithmProvider(null);
+          break;
+      }
+
+      // Randomly call method from the JWE algorithm provider instance
+      if (data.consumeBoolean()) {
+        JWEEncryptionProvider provider = new AesGcmJWEEncryptionProvider(data.pickValue(enc));
```
Comment (AdamKorcz) [inline projects/keycloak/JweAlgorithmProviderFuzzer.java:96]: Does the fuzzer need to create a new provider on every iteration, or can it be taken out of the fuzzers body and reused when needed?

**Review by AdamKorcz — CHANGES_REQUESTED** [review]

File: projects/keycloak/CertificateUtilsProviderFuzzer.java (hunk 1 above)
Comment (arthurscchan) [reply projects/keycloak/CertificateUtilsProviderFuzzer.java:106]: Comment added.

File: projects/keycloak/ECDSACryptoProviderFuzzer.java (hunk 2 above)
Comment (arthurscchan) [reply projects/keycloak/ECDSACryptoProviderFuzzer.java]: Removed.

File: projects/keycloak/JweAlgorithmProviderFuzzer.java (hunk 3 above)
Comment (arthurscchan) [reply projects/keycloak/JweAlgorithmProviderFuzzer.java]: Comment changed.

File: projects/keycloak/JweAlgorithmProviderFuzzer.java (hunk 4 above)
Comment (arthurscchan) [reply projects/keycloak/JweAlgorithmProviderFuzzer.java:132]: Comment changed.

File: projects/keycloak/build.sh (hunk 5 above)
Comment (arthurscchan) [reply projects/keycloak/build.sh:32]: Removed when build.sh is changed in #422.

File: projects/keycloak/build.sh (hunk 6 above)
Comment (arthurscchan) [reply projects/keycloak/build.sh:42]: Logic changed for build.sh in #422.

File: projects/keycloak/JweAlgorithmProviderFuzzer.java (hunk 7 above)
Comment (arthurscchan) [reply projects/keycloak/JweAlgorithmProviderFuzzer.java:58]: Logic changed.

File: projects/keycloak/JweAlgorithmProviderFuzzer.java (hunk 8 above)
Comment (arthurscchan) [reply projects/keycloak/JweAlgorithmProviderFuzzer.java:83]: Fixed.

File: projects/keycloak/CertificateUtilsProviderFuzzer.java (hunk 9 above)
Comment (arthurscchan) [reply projects/keycloak/CertificateUtilsProviderFuzzer.java]: fixed.

File: projects/keycloak/ECDSACryptoProviderFuzzer.java (hunk 10 above)
Comment (arthurscchan) [reply projects/keycloak/ECDSACryptoProviderFuzzer.java]: fixed.

File: projects/keycloak/JweAlgorithmProviderFuzzer.java (hunk 11 above)
Comment (arthurscchan) [reply projects/keycloak/JweAlgorithmProviderFuzzer.java]: Fixed.

File: projects/keycloak/ECDSACryptoProviderFuzzer.java (hunk 12 above)
Comment (arthurscchan) [reply projects/keycloak/ECDSACryptoProviderFuzzer.java]: Fixed.

**Comment by arthurscchan** [comment]
This PR needs rebase after #422 is merged in.

**Comment by AdamKorcz** [comment]
> This PR needs rebase after #422 is merged in.

Can be rebased.

File: projects/keycloak/CertificateUtilsProviderFuzzer.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        case 5:
+          keyPair = KeyUtils.generateRsaKeyPair(2048);
+          Date startDate = new Date(data.consumeLong());
+          Date expiryDate = new Date(data.consumeLong());
+          provider.createServicesTestCertificate(data.consumeString(data.remainingBytes() / 2),
+              startDate, expiryDate, keyPair, data.consumeRemainingAsString());
+          break;
+      }
+    } catch (Exception e) {
+      // Known exception directly thrown from the above methods
```
Comment (AdamKorcz) [inline projects/keycloak/CertificateUtilsProviderFuzzer.java:107]: Please elaborate in the comment why we don't catch specific exceptions.

File: projects/keycloak/JweAlgorithmProviderFuzzer.java [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+        storage.setEncryptionProvider(provider);
+        storage.setEncryptionKey(encryptionKey);
+        storage.setDecryptionKey(decryptionKey);
+
+        algorithmProvider.encodeCek(provider, storage, encryptionKey);
+      } else {
+        algorithmProvider.decodeCek(data.consumeRemainingAsBytes(), decryptionKey);
+      }
+    } catch (Exception e) {
+      // Known exception thrown directly from the encode or decode method.
```
Comment (AdamKorcz) [inline projects/keycloak/JweAlgorithmProviderFuzzer.java:133]: Please elaborate in the comment why we don't catch specific exceptions.

File: projects/keycloak/CertificateUtilsProviderFuzzer.java [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+          break;
+      }
+
+      // Randomly choose which method to invoke
+      Integer choice = data.consumeInt(1, 4);
+      switch (choice) {
+        case 1:
+          cert = (X509Certificate) cf.generateCertificate(
+              new ByteArrayInputStream(data.consumeBytes(data.remainingBytes() / 2)));
+          keyPair = KeyUtils.generateRsaKeyPair(2048);
```
Comment (AdamKorcz) [inline projects/keycloak/CertificateUtilsProviderFuzzer.java:80]: Please move all instances of `generateRsaKeyPair` out of the fuzzer body and reuse. It looks like there is no value from generating a new one every time.

**Comment by arthurscchan** [comment]
For JweAlgorithmProviderFuzzer and CertificateUtilsProviderFuzzer, the code have some initialization logic in the constructor, thus the provider constructor needed to be called for each fuzzing.

File: projects/keycloak/CertificateUtilsProviderFuzzer.java [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+        case 2:
+            ecuProvider = new ElytronCertificateUtils();
+          break;
+        case 3:
+            bfcuProvider = new BCFIPSCertificateUtilsProvider();
+          break;
+      }
+
+      // Randomly choose which method to invoke
+      Integer choice = data.consumeInt(1, 4);
```
Comment (AdamKorcz) [inline projects/keycloak/CertificateUtilsProviderFuzzer.java:74]: Should be up to 5, right?

File: projects/keycloak/CertificateUtilsProviderFuzzer.java [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+  public static void fuzzerInitialize() {
+    try {
+      // Initialize certificate factory
+      cf = CertificateFactory.getInstance("X.509");
+
+      // Initialize key pair
+      KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA");
+      generator.initialize(2048);
+      keyPair = generator.generateKeyPair();
+    } catch (CertificateException | NoSuchAlgorithmException e) {
```
Comment (AdamKorcz) [inline projects/keycloak/CertificateUtilsProviderFuzzer.java:49]: I think we should terminate the fuzzer if this fails. If this fails, then there is no point in letting the fuzzer run.

**Review by AdamKorcz — CHANGES_REQUESTED** [review]

**Review by AdamKorcz — APPROVED** [review]

**Pull request merged**